### PR TITLE
React/advanced pvf

### PR DESCRIPTION
**Pull request by kzalite**

### CODE REVIEW

File: app/ts/PreferencesTab/Preferences/PartialValueFunctions/AdvancedPartialValueFunctions/AdvancedPvfButtons/AdvancedPvfButtons.tsx [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+import {AdvancedPartialValueFunctionContext} from '../AdvancedPartialValueFunctionContext/AdvancedPartialValueFunctionContext';
+
+export default function AdvancedPvfButtons(): JSX.Element {
+  const {setActiveView, setPieceWisePvf} = useContext(PreferencesContext);
+  const {configuredRanges} = useContext(SubproblemContext);
+  const {advancedPvfCriterion, direction, cutOffs, isSaveDisabled} = useContext(
+    AdvancedPartialValueFunctionContext
+  );
+
+  const configuredRange =
```
Comment (kzalite) [inline app/ts/PreferencesTab/Preferences/PartialValueFunctions/AdvancedPartialValueFunctions/AdvancedPvfButtons/AdvancedPvfButtons.tsx:17]: Should we make a getConfiguredRange(criterion) for Subproblem context? This code many places, such repeat.

File: app/ts/PreferencesTab/Preferences/PartialValueFunctions/AdvancedPartialValueFunctions/PvfDirection/PvfDirection.tsx [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+      </Grid>
+      <Grid item xs={12}>
+        <RadioGroup
+          name="pvf-direction-radio"
+          value={direction}
+          onChange={handleRadioChanged}
+        >
+          <label id="increasing-pvf-option">
+            <Radio value="increasing" /> Increasing (
+            {getPercentifiedValue(configuredRange[1], usePercentage)} is best)
```
Comment (kzalite) [inline app/ts/PreferencesTab/Preferences/PartialValueFunctions/AdvancedPartialValueFunctions/PvfDirection/PvfDirection.tsx:41]: Should we add unit?

File: app/ts/PreferencesTab/Preferences/PartialValueFunctions/PartialValueFunctionButtons/PartialValueFunctionButtons.tsx [hunk 3]
```diff
@@ -58,7 +57,7 @@ export default function PartialValueFunctionButtons({
           Decreasing
         </Button>
       </Tooltip>
-      {/* <Tooltip title="Set linear or piece-wise PVF via guided process. Setting a PVF will reset all trade-off preferences.">
+      <Tooltip title="Set piece-wise PVF via guided process. Setting a PVF will reset all trade-off preferences.">
```
Comment (kzalite) [inline app/ts/PreferencesTab/Preferences/PartialValueFunctions/PartialValueFunctionButtons/PartialValueFunctionButtons.tsx:60]: piecewise?

File: app/ts/PreferencesTab/Preferences/PartialValueFunctions/PartialValueFunctionUtil.ts [hunk 4]
```diff
@@ -21,53 +23,46 @@ function getXValues(pvf: IPvf, usePercentage: boolean): ['x', ...number[]] {
...
+function intermediateY(pvf: TPvf): number[] {
+  if (isPieceWiseLinearPvf(pvf)) {
+    return pvf.values;
+  } else {
+    return [];
+  }
 }
 
-export function getBest(pvf: IPvf, usePercentage: boolean): number {
+export function getBest(pvf: TPvf, usePercentage: boolean): number {
```
Comment (kzalite) [inline app/ts/PreferencesTab/Preferences/PartialValueFunctions/PartialValueFunctionUtil.ts:51]: Throw if no 'direction'?

File: app/ts/PreferencesTab/PreferencesContext.tsx [hunk 5]
```diff
@@ -86,51 +90,45 @@ export function PreferencesContextProviderComponent({
...
+    );
     updateScenario(newScenario).then(() => {
       if (areAllPvfsSet(newPvfs)) {
         resetPreferences(newScenario);
       }
     });
   }
 
-  function areAllPvfsSet(newPvfs: Record<string, IPvf>): boolean {
+  function areAllPvfsSet(newPvfs: Record<string, TPvf>): boolean {
```
Comment (kzalite) [inline app/ts/PreferencesTab/PreferencesContext.tsx:119]: to util + test

File: app/ts/PreferencesTab/PreferencesContext.tsx [hunk 6]
```diff
@@ -225,13 +223,34 @@ export function PreferencesContextProviderComponent({
...
   }
 
-  function getCriterion(id: string): ICriterion {
-    return _.find(filteredCriteria, ['id', id]);
+  function goToAdvancedPvf(criterionId: string): void {
+    setAdvancedPvfCriterionId(criterionId);
+    setActiveView('advancedPvf');
+  }
+
+  function setPieceWisePvf(
```
Comment (kzalite) [inline app/ts/PreferencesTab/PreferencesContext.tsx:231]: can be combined with setLinearPvf?

File: app/ts/PreferencesTab/PreferencesUtil.test.ts [hunk 7]
```diff
@@ -80,7 +80,7 @@ const criterion3: ICriterion = {
 
 describe('PreferencesUtil', () => {
   describe('initPvfs', () => {
-    const criteria: ICriterion[] = [criterion1, criterion2];
+    const criteria: ICriterion[] = [criterion1];
```
Comment (kzalite) [inline app/ts/PreferencesTab/PreferencesUtil.test.ts:83]: test case for a criterion with no scenario pvfs

Comment (kzalite) [reply app/ts/PreferencesTab/PreferencesUtil.test.ts:83]: MORE test cases, now with piecewise linear things

File: app/ts/util/PataviUtil.ts [hunk 8]
```diff
@@ -24,7 +24,7 @@ import {
   TPerformance
 } from '@shared/interface/Problem/IPerformance';
 import IProblem from '@shared/interface/Problem/IProblem';
-import IPvf from '@shared/interface/Problem/IPvf';
+import IPvf, {TPvf} from '@shared/interface/Problem/IPvf';
```
Comment (kzalite) [inline app/ts/util/PataviUtil.ts:27]: IPvf should not be used

File: app/ts/PreferencesTab/Preferences/PartialValueFunctions/AdvancedPartialValueFunctions/AdvancedPvfButtons/AdvancedPvfButtons.tsx (hunk 1 above)
Comment (DanielReid) [reply app/ts/PreferencesTab/Preferences/PartialValueFunctions/AdvancedPartialValueFunctions/AdvancedPvfButtons/AdvancedPvfButtons.tsx:17]: makes sense

File: app/ts/PreferencesTab/Preferences/PartialValueFunctions/AdvancedPartialValueFunctions/CutOffs/CutOffs.tsx [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+    max: significantDigits(configuredRange[1] - stepSize),
+    values: cutOffs,
+    formatFunction: (x: number) => getPercentifiedValue(x, usePercentage)
+  };
+
+  function handleSliderChanged(
+    event: React.ChangeEvent<any>,
+    newValue: [number, number, number]
+  ) {
+    setCutOffs(_.sortBy(newValue) as [number, number, number]);
```
Comment (DanielReid) [inline app/ts/PreferencesTab/Preferences/PartialValueFunctions/AdvancedPartialValueFunctions/CutOffs/CutOffs.tsx:39]: oh, that's a shame that `as` is necessary here

File: app/ts/PreferencesTab/Preferences/PartialValueFunctions/AdvancedPartialValueFunctions/PvfDirection/PvfDirection.tsx (hunk 2 above)
Comment (DanielReid) [reply app/ts/PreferencesTab/Preferences/PartialValueFunctions/AdvancedPartialValueFunctions/PvfDirection/PvfDirection.tsx:41]: shhhhhhh

File: app/ts/PreferencesTab/Preferences/PartialValueFunctions/PartialValueFunctionUtil.ts (hunk 4 above)
Comment (DanielReid) [reply app/ts/PreferencesTab/Preferences/PartialValueFunctions/PartialValueFunctionUtil.ts:51]: if we pass `TPvf` in the direction should already be set right? so the issue should happen earlier in the chain

File: shared/interface/Pvfs/ILinearPvf.ts [hunk 10]
```diff
@@ -0,0 +1,7 @@
+import {TPvfDirection} from '@shared/types/TPvfDirection';
+import IPvf from '../Problem/IPvf';
+
+export interface ILinearPvf extends IPvf {
+  type: 'linear';
+  direction: TPvfDirection;
```
Comment (DanielReid) [inline shared/interface/Pvfs/ILinearPvf.ts:6]: it's kind of weird to have both interfaces that extend `IPvf` both have an identical `direction` property, indicating their base shares that as well

File: app/ts/PreferencesTab/PreferencesUtil.ts [hunk 11]
```diff
@@ -87,3 +123,34 @@ export function determineElicitationMethod(preferences: TPreferences): string {
...
+  pvf: TPvf,
+  currentScenario: IMcdaScenario
+) {
+  let newScenario: IMcdaScenario = _.cloneDeep(currentScenario);
+  if (!newScenario.state.problem) {
+    newScenario.state.problem = {criteria: {}};
+  }
+  if (isPieceWiseLinearPvf(pvf)) {
+    newScenario.state.problem.criteria[criterionId] = {
+      dataSources: [
```
Comment (DanielReid) [inline app/ts/PreferencesTab/PreferencesUtil.ts:138]: the dataSources step in here is so weird

File: app/ts/util/PataviUtil.ts (hunk 8 above)
Comment (DanielReid) [reply app/ts/util/PataviUtil.ts:27]: could consider not even exporting it and putting the other interface declarations that extend it in the same file instead

**Review by DanielReid — CHANGES_REQUESTED** [review]
Very neat. Some missing changes as said

**Comment by kzalite** [comment]
omg, all tests passed

File: app/ts/PreferencesTab/Preferences/PartialValueFunctions/AdvancedPartialValueFunctions/AdvancedPvfPlot/AdvancedPvfPlot.tsx [hunk 12]
```diff
@@ -0,0 +1,49 @@
...
+
+  const width = '500px';
+  const height = '400px';
+
+  useEffect(() => {
+    const values =
+      direction === 'decreasing'
+        ? [1, 0.75, 0.5, 0.25, 0]
+        : [0, 0.25, 0.5, 0.75, 1];
+    const settings: ChartConfiguration = generateAdvancedPlotSettings(
```
Comment (chinook25) [inline app/ts/PreferencesTab/Preferences/PartialValueFunctions/AdvancedPartialValueFunctions/AdvancedPvfPlot/AdvancedPvfPlot.tsx:28]: generateAdvancedPlotSettings is to generic of a name, generateAdvancedPvfPlotSettings would be more appropriate.

File: app/ts/PreferencesTab/PreferencesContext.tsx [hunk 13]
```diff
@@ -225,13 +188,41 @@ export function PreferencesContextProviderComponent({
...
 
-  function getCriterion(id: string): ICriterion {
-    return _.find(filteredCriteria, ['id', id]);
+  function goToAdvancedPvf(criterionId: string): void {
+    setAdvancedPvfCriterionId(criterionId);
+    setActiveView('advancedPvf');
+  }
+
+  function setPvf(criterionId: string, pvf: TPvf): void {
+    const newPvfs = {..._.cloneDeep(pvfs), [criterionId]: pvf};
```
Comment (chinook25) [inline app/ts/PreferencesTab/PreferencesContext.tsx:197]: cloneDeep necessary?

File: app/ts/Settings/SettingsContext.tsx [hunk 14]
```diff
@@ -106,6 +115,7 @@ export function SettingsContextProviderComponent({children}: {children: any}) {
         settings,
         showPercentages: settings.showPercentages === 'percentage',
```
Comment (chinook25) [inline app/ts/Settings/SettingsContext.tsx:116]: With the new function, is this still needed?

File: test/endToEnd/partialValueFunctionTest.js [hunk 15]
```diff
@@ -3,13 +3,11 @@
 module.exports = {
   beforeEach: beforeEach,
   afterEach: afterEach,
-  // 'Set partial value functions': set,
-  // 'Go to previous step': navigate,
+  'Set piecewise linear partial value function': set,
```
Comment (chinook25) [inline test/endToEnd/partialValueFunctionTest.js:6]: should this test also fiddle with the slider a bit?

**Review by chinook25 — COMMENTED** [review]
nice

**Pull request merged**